src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpAssemblyVisitor.cs

```diff
@@ -0,0 +1,362 @@
+                                 bool hideImplicitDefaultConstructors = true)
+    {
+        _logger = logger;
+        _loader = loader;
+        _symbolFilter = symbolFilter;
+        _attributeDataSymbolFilter = attributeDataSymbolFilter;
+        _exceptionMessage = exceptionMessage;
+        _includeAssemblyAttributes = includeAssemblyAttributes;
+        _adhocWorkspace = new AdhocWorkspace();
+        _syntaxGenerator = SyntaxGenerator.GetGenerator(_adhocWorkspace, LanguageNames.CSharp);
+        _metadataReferences = metadataReferences;
+        _addPartialModifier = addPartialModifier;
+        _hideImplicitDefaultConstructors = hideImplicitDefaultConstructors;
+    }
+
+    /// <inheritdoc />
+    public Document GetDocumentForAssembly(IAssemblySymbol assemblySymbol)
+    {
+        CSharpCompilationOptions compilationOptions = new(OutputKind.DynamicallyLinkedLibrary,
+               nullableContextOptions: NullableContextOptions.Enable);
+        Project project = _adhocWorkspace.AddProject(ProjectInfo.Create(
+            ProjectId.CreateNewId(), VersionStamp.Create(), assemblySymbol.Name, assemblySymbol.Name, LanguageNames.CSharp,
+            compilationOptions: compilationOptions));
+        project = project.AddMetadataReferences(_metadataReferences ?? _loader.MetadataReferences);
+
+        IEnumerable<INamespaceSymbol> namespaceSymbols = EnumerateNamespaces(assemblySymbol).Where(_symbolFilter.Include);
+        List<SyntaxNode> namespaceSyntaxNodes = [];
+
+        foreach (INamespaceSymbol namespaceSymbol in namespaceSymbols.Order())
+        {
+            SyntaxNode? syntaxNode = Visit(namespaceSymbol);
+
+            if (syntaxNode is not null)
+            {
+                namespaceSyntaxNodes.Add(syntaxNode);
+            }
+        }
+
+        SyntaxNode compilationUnit = _syntaxGenerator.CompilationUnit(namespaceSyntaxNodes)
+            .WithAdditionalAnnotations(Formatter.Annotation, Simplifier.Annotation)
+            .Rewrite(new TypeDeclarationCSharpSyntaxRewriter(_addPartialModifier))
+            .Rewrite(new BodyBlockCSharpSyntaxRewriter(_exceptionMessage));
+
+        if (_includeAssemblyAttributes)
+        {
+            compilationUnit = GenerateAssemblyAttributes(assemblySymbol, compilationUnit);
+        }
+
+        compilationUnit = GenerateForwardedTypeAssemblyAttributes(assemblySymbol, compilationUnit);
+        compilationUnit = compilationUnit.NormalizeWhitespace(eol: Environment.NewLine);
+
+        Document document = project.AddDocument(assemblySymbol.Name, compilationUnit);
+        document = Simplifier.ReduceAsync(document).Result;
+        document = Formatter.FormatAsync(document, DefineFormattingOptions()).Result;
+
+        return document;
+    }
+
+    /// <inheritdoc />
+    public SyntaxNode GetFormattedRootNodeForDocument(Document document) => document.GetSyntaxRootAsync().Result!.Rewrite(new SingleLineStatementCSharpSyntaxRewriter());
```

Comment:
Does this actually use any instance state, or is it just a transformation on Document?  Why does this live on the assembly visitor?

Reply:
I renamed the class as suggested so the purpose of this method made more sense.

Reply:
Just tossing the idea out there - now that you have both CSharpAssemblyDocumentGenerator and CSharpFileBuilder do you think it makes sense to have these two different types?  Could we combine them back together and just have one that is used slightly differently depending on the caller?  It might reduce the total number of names we have to come up with while still being meaningful and keeping the diff smaller.